ehrql/file_formats/arrow.py

```diff
@@ -204,8 +204,11 @@ def _validate_column(self, name, column, spec):
                 f"Type mismatch in column '{name}': "
                 f"expected {spec.type}, got {column.type}"
             )
-        if pyarrow.types.is_dictionary(column.type) and spec.categories is not None:
-            column_categories = column.dictionary.to_pylist()
+        if spec.categories is not None:
+            if pyarrow.types.is_dictionary(column.type):
+                column_categories = column.dictionary.to_pylist()
+            else:
+                column_categories = [v for v in column.to_pylist() if v is not None]
```

Comment:
Won't `column_categories` contain duplicate values? Also, what would happen if the categories were in different batches?

Reply:
> Won't `column_categories` contain duplicate values?

It's setted on the next line.

> Also, what would happen if the categories were in different batches?

It doesn't matter -- it's only a problem if the column contains values that are not in the column spec, not vice versa.

Reply:
> It's setted on the next line.

Ah, I see. Thank you! But don't we use a list of `column_categories`, which may contain duplicate values, when returning the string? (I may have misunderstood what `column` contains.)

> It doesn't matter -- it's only a problem if the column contains values that are not in the column spec, not vice versa.

We validate only the first batch, don't we? So if a column in the second batch contained values that were not in the column spec, then we wouldn't know, would we?